[chore] System Semantic Conventions Non-Normative Guidance (#1618)
open-telemetry/semantic-conventions: docs/non-normative/groups/system/use-cases.md

# **System Semantic Conventions: General Use Cases**

This document is a collection of the use cases that we want to cover with the System Semantic Conventions. The use cases outlined here inform the working group’s decisions around what instrumentation is considered **required**.  
Use cases in this document will be stated in a generic way that does not refer to any potentially existing instrumentation in semconv as of writing, such that when we do dig into specific instrumentation, we understand their importance based on our holistic view of expected use cases.

## *Legend*

`General Information` \= The information that should be discoverable either through the entity, metrics, or metric attributes.

`Dashboard` \= The information that should be attainable through metrics to create a comprehensive dashboard.

`Alerts` \= Some examples of common alerts that should be creatable with the available information.

## **Host**

> A user should be able to monitor the health of a host, including monitoring resource consumption, unexpected errors due to resource exhaustion or malfunction of core components of a host or fleet of hosts (network stack, memory, CPU…).

Reviewer: > unexpected errors due to resource exhaustion

not sure if we have anything defined today and if there is anything general we can provide, but it'd be nice to have some OS network/hw/etc errors and have them on the dashboards/alerts

Reviewer 2: We have the [`system.network.errors`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/system-metrics.md#metric-systemnetworkerrors) metric, I don't think we have anything else (I don't know if there is a way to retrieve this, libraries like psutil don't provide this for other stuff like memory or disk AFAIK). Still, I think the existing metrics cover the case of troubleshooting resource exhaustion/malfunction

### General Information

* Machine name  
* ID (relevant to its context, could be a cloud provider ID or just base machine ID)  
* OS information (platform, version, architecture, etc)  

> * Number of CPU cores  

Reviewer 2: Maybe this can be "CPU information" instead? We have a bunch of those [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/host.md)

* Memory Capacity

### Dashboard

* Memory utilization  

> * CPU utilization  

Reviewer: do we support or track GPU metrics anywhere? (totally fine to leave it out of scope of this doc)

Reviewer 2: The `hw.` metrics have some of it https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/hardware-metrics.md#hwgpu---gpu-metrics. We haven't really addressed these much.

* Disk utilization  
* Disk throughput  
* Network traffic

### Alerts

> * VM is up  

Reviewer: ```suggestion
* VM is down unexpectedly
```

?

Author: Done in [487af83](https://github.com/open-telemetry/semantic-conventions/pull/1618/commits/487af83f99a9ab8eeeb937e6cfac28638114cd8b)

* Memory/CPU/Disk utilization goes above a % threshold  
* Network activity spikes unexpectedly

## Notes

* The alerts in particular should be capable of being uniformly applied to an entire fleet of hosts  
* The user may be monitoring a virtualization host i.e. VMWare or Proxmox, and the instrumentation to monitor the health of the root host and the virtual machines it's spawned can be largely the same

## **Process**

A user should be able to monitor the health of an arbitrary process using data provided by the OS.  
Reasons a user may want this:

> 1. The process they want to monitor isn't covered by more specific semconv instrumentation such as language runtime metrics, db, http, etc.  

Reviewer: maybe something along these lines?

```suggestion
1. The process they want to monitor doesn't have in-process runtime-specific instrumentation enabled or is not instrumentable at all such as antivirus or another background process.  
```

i.e. what matters seems to be whether you have in-proc instrumentation and if you can apply it at all.

Author: Done in [487af83](https://github.com/open-telemetry/semantic-conventions/pull/1618/commits/487af83f99a9ab8eeeb937e6cfac28638114cd8b)

2. They are monitoring lots of processes and want to have a set of uniform instrumentation for all of them.  
3. Personal preference/legacy reasons; they might already be using OS signals to monitor stuff and it's an easier lift for them to move to basic process instrumentation, then move to other specific semconv over time.

### General Information

* Process name  
* Pid  
* User/owner

### Dashboard

* Physical Memory usage and/or utilization  
* Virtual Memory usage  
* CPU usage and/or utilization  
* Disk throughput  
* Network throughput

### Alert

> * Process stops  

Reviewer: nit

```suggestion
* Process stops unexpectedly
```

* Memory/CPU usage/utilization goes above a threshold  
* Memory exclusively rises over a period of time (memory leak detection)

### Notes

* Unless the OS provides the utilization data directly, the utilization requires calculation. Process instrumentation would need to be associated with a host entity that contains data about its memory capacity for utilization metrics to be calculated.  

> * Process instrumentation can also be used as data for benchmark evaluations, collecting the data for a period of time and evaluating the timeseries to get benchmarking/overhead insights about the process

Reviewer: is the intent here to say that process resource utilization metrics are useful for all sorts of perf testing?

Author: Reworded in [487af83](https://github.com/open-telemetry/semantic-conventions/pull/1618/commits/487af83f99a9ab8eeeb937e6cfac28638114cd8b)
